Payment Checkout: implement redirect modal in the checkout page for Agency/Pro users from Jetpack Pricing or My Jetpack page

yashwin opened the pull request:
Related to 1202619025189113-as-1204953498074085

## Proposed Changes

This PR implements a modal in the payment checkout page that shows the Agency/Pro users the benefits of buying the products from the Jetpack Licensing Portal. 

We show this page only if 

- The user is an agency partner.
- If the user is coming from Jetpack Pricing or My Jetpack page.

#### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/show-redirect-modal-in-checkout` and `yarn start` or open the Calypso live link. 
2. Open http://calypso.localhost:3000/ or the Calypso live link.
3. Visit https://cloud.jetpack.com/pricing and click on the `Get` button on any product -> You'll be redirected to the payment checkout page -> Replace `https://wordpress.com/` with `http://calypso.localhost:3000/ ` or the live link URL.
4. Verify that you can see the modal as shown below

![mobile (23)](https://github.com/Automattic/wp-calypso/assets/10586875/a79f2eb2-036b-4aa9-898b-00ea0ddacfca)

5. Go to WP.com -> Select any Jetpack site -> Visit WP Admin page -> Click on Jetpack - My Jetpack -> Click on `Start for free` on the Search card -> Click the `Get Jetpack Search` button -> Replace `https://wordpress.com/` with `http://calypso.localhost:3000/ ` or the live link URL and verify you can now see the same modal.

6. Go to WP.com -> Select any Jetpack site -> Visit WP Admin page -> Click on Jetpack - My Jetpack -> Click on `Purchase` on any product(Backup or Scan) -> Click the get product button -> Replace `https://wordpress.com/` with `http://calypso.localhost:3000/ ` or the live link URL and verify you can now see the same modal.
7. Verify the modal looks good on different screen sizes

**Tablet**

![mobile (24)](https://github.com/Automattic/wp-calypso/assets/10586875/4a90923b-4ff0-4ef6-b06b-801afe4cb120)

**Mobile**

![mobile (25)](https://github.com/Automattic/wp-calypso/assets/10586875/e7e8a569-743b-482f-bd25-4939382f81ac)

8. Go to WP.com and add any WP.com plans(Personal or Business) by clicking the Upgrade button and verify that the modal is not shown now.

<img width="1266" alt="Screenshot 2023-07-05 at 10 11 14 AM" src="https://github.com/Automattic/wp-calypso/assets/10586875/e18c81b6-03f6-4b4c-b1c0-7dd2c725a7d6">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yashwin committed:
Implement redirect modal in the checkout page

simonktnga8c commented:
Just tested and it works great; nice work @yashwin!

ilonagl reviewed (approved):
Looks great. Amazing work!

sirbrillig commented on client/my-sites/checkout/composite-checkout/components/checkout-main.tsx:
```diff
@@ -722,26 +728,34 @@ export default function CheckoutMain( {
-					siteId={ updatedSiteId }
-					siteUrl={ updatedSiteSlug }
-				/>
+				<>
```
In theory, you should be able to have multiple sibling children of `CheckoutProvider` without needing to add this fragment, but I'm guessing you did it because there was a problem. If that's so, what was the issue that this solved?

yashwin replied:
Yeah. It did work - https://github.com/Automattic/wp-calypso/pull/78955/commits/6ebd952eeb69ec8e94332682268322499ad30496

sirbrillig commented on client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx:
```diff
@@ -0,0 +1,111 @@
+import { Button } from '@automattic/components';
+import { Modal } from '@wordpress/components';
+import { useTranslate } from 'i18n-calypso';
+import { CheckIcon } from 'calypso/../packages/composite-checkout/src/components/shared-icons';
```
This should be a package import. We should not reach inside the private implementation details of a package. 

```suggestion
import { CheckIcon } from '@automattic/composite-checkout';
```

yashwin replied:
I think this happened because of auto-import. I agree with this and fixed it here - https://github.com/Automattic/wp-calypso/pull/78955/commits/6ebd952eeb69ec8e94332682268322499ad30496

sirbrillig replied:
Yeah, that happens to me all the time too. I should check with the Calypso team to see if there's a way to add a rule to catch these sorts of imports. 🤔 @jsnajdr do you think that would be possible and/or would it be too much work?

jsnajdr replied:
> see if there's a way to add a rule to catch these sorts of imports.

You mean an eslint rule? Maybe `eslint-plugin-import` has some rule for this out of the box. If we wanted to write our own rule, it shouldn't be terribly hard. We'd mainly have to specify what exactly are the imports we want to avoid.

The auto-import suggestions are created by VSCode, which in turn uses the JS Language Server to generate them.

sirbrillig replied:
>  We'd mainly have to specify what exactly are the imports we want to avoid.

I think that it would be safe to say that we want to block all imports that try to reach a path starting with `calypso/../packages`. To also block relative imports we could say any path including `../packages` but it is possible that could find false positives.

sirbrillig commented on client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx:
```diff
@@ -0,0 +1,111 @@
+						</p>
+						<ul className="jetpack-pro-redirect-modal__list">
+							{ features.map( ( feature, index ) => (
+								<li key={ index }>
```
[You should not use the index as the key](https://react.dev/learn/rendering-lists#why-does-react-need-keys). There's a ton of other articles about why this is dangerous out there on the web if you need additional info. Can we use `feature` here?

yashwin replied:
Totally agree, and sorry for missing this. I have fixed it now - https://github.com/Automattic/wp-calypso/pull/78955/commits/6ebd952eeb69ec8e94332682268322499ad30496

sirbrillig reviewed (changes requested):
Thank you for providing screenshots at different widths! Just a few comments.

yashwin committed:
Address PR comments

yashwin commented:
@sirbrillig: Thanks for your thorough review and feedback. I have addressed all your comments. Could you please take another look?

sirbrillig reviewed (approved):
I haven't run through the testing instructions but from a payments perspective, this looks good to me. Thanks for the fixes!

jkguidaven commented on client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx:
```diff
@@ -0,0 +1,114 @@
+			[ 'my-jetpack', 'jetpack' ].includes( redirectURLPage )
+		)
+	) {
+		return null;
```
Not a blocker.

I see two separate places where we check if we need to display this modal. What do you think about moving this logic to the `checkout-main.tsx` or adding the Agency user check here?

jkguidaven commented on client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	return (
+		<>
```
No need to wrap the Modal with `<>`.

jkguidaven commented on client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/style.scss:
```diff
@@ -0,0 +1,124 @@
+		padding: 16px 32px;
+		background: var(--studio-white);
+
+		.jetpack-pro-redirect-modal__footer-buttons {
```
Just nitpicking but I think we can shorten this to `&-buttons`.

yashwin replied:
I would like to keep it like this, keeping searchability in mind.

jkguidaven commented on client/state/jetpack-agency-dashboard/selectors.ts:
```diff
@@ -16,6 +16,9 @@ export const JETPACK_DASHBOARD_WELCOME_BANNER_PREFERENCE_HOME_PAGE =
 	'jetpack-dashboard-agency-program-survey-banner-preference';
 
+export const JETPACK_DASHBOARD_CHECKOUT_REDIRECT_MODAL_DISMISSED =
+	'jetpack-dashboard-checkout-redirect-modal-dismissed';
```
Not a blocker.

Maybe we can add the word `agency-program` in the preference name just to make it more specific to the agency program and not overlap with the Jetpack dashboard in the core plugin. wdyt?

yashwin replied:
Doesn't really matter. But, I have changed this.

jkguidaven commented on client/my-sites/checkout/composite-checkout/components/jetpack-pro-redirect-modal/index.tsx:
```diff
@@ -0,0 +1,114 @@
+			<Modal
+				open={ true }
+				onRequestClose={ dismissAndRecordEvent }
+				title=""
```
Maybe what we need here is **Dialog** instead of **Modal**.
```
import { Dialog } from '@automattic/components';
```

yashwin replied:
I'm not going to make this change here as it requires a lot of changes.Also, there is not much difference between using these 2 components.

jkguidaven reviewed (commented):
Tested it and it is working as expected. Nice work @yashwin 💯 👍 

However, I left few comments.

yashwin committed:
Address PR comments

yashwin commented:
@jkguidaven - Thanks for your thorough review. I have addressed all your comments and made the changes - https://github.com/Automattic/wp-calypso/pull/78955/commits/7be78d89fc5eaf2cc2f65f6493f323ce01141012

We would want to verify this for non-agency users, as we have moved the conditions to one place. This modal should ideally not appear to non-agency users.

jkguidaven commented:
Thanks for addressing all my PR comments. test both agency and non-agency user flow. this LGTM 👍 🚢

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8359905

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8359905).
